src/main/java/io/cdap/plugin/couchbase/source/N1qlQueryRowRecordReader.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * RecordReader implementation, which reads N1qlQueryRow entries.
+ */
+public class N1qlQueryRowRecordReader extends RecordReader<NullWritable, N1qlQueryRow> {
+  private static final Logger LOG = LoggerFactory.getLogger(N1qlQueryRowRecordReader.class);
+  private static final Gson gson = new GsonBuilder().create();
+
+  private Cluster cluster;
+  private Bucket bucket;
+  private Iterator<N1qlQueryRow> iterator;
+  private N1qlQueryRow value;
+
+  /**
+   * Initialize an iterator and config.
+   *
+   * @param inputSplit         specifies batch details
+   * @param taskAttemptContext task context
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) {
+    Configuration conf = taskAttemptContext.getConfiguration();
+    String configJson = conf.get(N1qlQueryRowInputFormatProvider.PROPERTY_CONFIG_JSON);
+    CouchbaseSourceConfig config = gson.fromJson(configJson, CouchbaseSourceConfig.class);
+
+    this.cluster = CouchbaseCluster.create(config.getNodeList());
+    if (!Strings.isNullOrEmpty(config.getUser()) || !Strings.isNullOrEmpty(config.getPassword())) {
+      cluster.authenticate(config.getUser(), config.getPassword());
+    }
+    this.bucket = cluster.openBucket(config.getBucket());
+
+    Statement statement = Strings.isNullOrEmpty(config.getConditions())
+      ? Select.select(config.getSelectFields()).from(Expression.i(config.getBucket()))
+      : Select.select(config.getSelectFields()).from(Expression.i(config.getBucket())).where(config.getConditions());
+    LOG.trace("Executing query split: {}", statement);
+
+    N1qlQuery query = N1qlQuery.simple(statement);
+    query.params().consistency(config.getScanConsistency().getScanConsistency())
+      .maxParallelism(config.getMaxParallelism())
+      .serverSideTimeout(config.getTimeout(), TimeUnit.SECONDS);
+
+    N1qlQueryResult result = bucket.query(query);
+    if (!result.finalSuccess()) {
+      String errorMessage = result.errors().stream()
+        .map(JsonObject::toString)
+        .collect(Collectors.joining("\n"));
+      throw new CouchbaseExecutionException(errorMessage);
+    }
+
+    this.iterator = result.rows();
+  }
+
+  @Override
+  public boolean nextKeyValue() {
+    if (!iterator.hasNext()) {
+      return false;
+    }
+    value = iterator.next();
+    return true;
+  }
+
+  @Override
+  public NullWritable getCurrentKey() {
+    return null;
+  }
+
+  @Override
+  public N1qlQueryRow getCurrentValue() {
+    return value;
+  }
+
+  @Override
+  public float getProgress() {
+    // progress is unknown
+    return 0.0f;
+  }
+
+  @Override
+  public void close() throws IOException {
+    LOG.trace("Closing Record reader");
+    this.bucket.close();
```

Comment:
check that the bucket is not null before closing, same with the cluster

Reply:
Done.